akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -22,6 +25,125 @@ object ORMap {
+    type T = DeltaOp
+    override def zero: DeltaReplicatedData
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def emptyWithPNCounterMapTag[A, B <: ReplicatedData]: ORMap[A, B] = new ORMap(ORSet.empty, Map.empty, zeroTag = PNCounterMapTag)
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def emptyWithORMultiMapTag[A, B <: ReplicatedData]: ORMap[A, B] = new ORMap(ORSet.empty, Map.empty, zeroTag = ORMultiMapTag)
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def emptyWithORMultiMapWithValueDeltasTag[A, B <: ReplicatedData]: ORMap[A, B] = new ORMap(ORSet.empty, Map.empty, zeroTag = ORMultiMapWithValueDeltasTag)
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def emptyWithLWWMapTag[A, B <: ReplicatedData]: ORMap[A, B] = new ORMap(ORSet.empty, Map.empty, zeroTag = LWWMapTag)
+
+  /**
+   * INTERNAL API
+   * Tags for ORMap.DeltaOp's, so that when the Replicator needs to re-create full value from delta,
+   * the right map type will be used
+   */
+  @InternalApi private[akka] trait ZeroTag {
+    def zero: DeltaReplicatedData
+    def value: Int
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] case object VanillaORMapTag extends ZeroTag {
+    override def zero: DeltaReplicatedData = ORMap.empty
+    override final val value: Int = 0
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] case object PNCounterMapTag extends ZeroTag {
+    override def zero: DeltaReplicatedData = PNCounterMap.empty
+    override final val value: Int = 1
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] case object ORMultiMapTag extends ZeroTag {
+    override def zero: DeltaReplicatedData = ORMultiMap.empty
+    override final val value: Int = 2
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] case object ORMultiMapWithValueDeltasTag extends ZeroTag {
+    override def zero: DeltaReplicatedData = ORMultiMap.emptyWithValueDeltas
+    override final val value: Int = 3
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] case object LWWMapTag extends ZeroTag {
+    override def zero: DeltaReplicatedData = LWWMap.empty
+    override final val value: Int = 4
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] sealed abstract class AtomicDeltaOp[A, B <: ReplicatedData] extends DeltaOp {
+    def underlying: ORSet.DeltaOp
+    def zeroTag: ZeroTag
+    override def zero: DeltaReplicatedData = zeroTag.zero
+
+    override def merge(that: DeltaOp): DeltaOp = that match {
+      case other: AtomicDeltaOp[A, B] ⇒ DeltaGroup(Vector(this, other))
+      case DeltaGroup(ops)            ⇒ DeltaGroup(this +: ops)
+    }
+  }
+
+  // PutDeltaOp contains ORSet delta and full value
+  /** INTERNAL API */
+  @InternalApi private[akka] final case class PutDeltaOp[A, B <: ReplicatedData](underlying: ORSet.DeltaOp, value: (A, B), zeroTag: ZeroTag = VanillaORMapTag) extends AtomicDeltaOp[A, B] {
```

Comment:
Agreed. When I was introducing those, I have also omitted the tag, exactly for the reason you are giving. Will fix.